File: backend/app/domain/models/dataset.py
from enum import Enum
from typing import Any, Dict, List, Optional, Union
from pydantic import BaseModel, Field
from datetime import datetime


class VariableMeasure(str, Enum):
    nominal = "nominal"
    ordinal = "ordinal"
    scale = "scale"


class VariableType(str, Enum):
    numeric = "numeric"
    string = "string"
    date = "date"


class VariableRole(str, Enum):
    input = "input"
    target = "target"
    both = "both"
    none = "none"
    partition = "partition"
    split = "split"


class ValueLabel(BaseModel):
    value: Union[float, int, str]
    label: str


class VariableMeta(BaseModel):
    name: str
    label: str = ""
    var_type: VariableType = VariableType.numeric
    width: int = 8
    decimals: int = 2
    value_labels: Dict[str, str] = Field(default_factory=dict)
    missing_values: List[Union[float, int, str]] = Field(default_factory=list)
    measure: VariableMeasure = VariableMeasure.scale
    role: VariableRole = VariableRole.input


class DatasetMeta(BaseModel):
    file_name: str
    n_cases: int
    n_vars: int
    variables: List[VariableMeta]
    encoding: str = "utf-8"


class DatasetSession(BaseModel):
    session_id: str
    meta: DatasetMeta
    created_at: datetime = Field(default_factory=datetime.utcnow)

File: frontend/src/types/dataset.ts
export type VariableMeasure = "nominal" | "ordinal" | "scale";
export type VariableType = "numeric" | "string" | "date";
export type VariableRole = "input" | "target" | "both" | "none" | "partition" | "split";

export interface ValueLabel {
  value: number | string;
  label: string;
}

export interface VariableMeta {
  name: string;
  label: string;
  var_type: VariableType;
  width: number;
  decimals: number;
  value_labels: Record<string, string>;
  missing_values: (number | string)[];
  measure: VariableMeasure;
  role: VariableRole;
}

export interface DatasetMeta {
  file_name: string;
  n_cases: number;
  n_vars: number;
  variables: VariableMeta[];
  encoding: string;
}

export interface DatasetSession {
  session_id: string;
  meta: DatasetMeta;
  created_at: string;
}

export type JobStatusEnum = "PENDING" | "PROGRESS" | "SUCCESS" | "FAILURE";

export interface JobStatus {
  job_id: string;
  status: JobStatusEnum;
  progress_msg?: string;
  result?: unknown;
  error?: string;
}

export interface DataPage {
  data: Record<string, unknown>[];
  total: number;
  page: number;
  page_size: number;
}

export interface OutputBlock {
  id: string;
  type: "table" | "chart";
  title: string;
  subtitle?: string;
  content: unknown;
  created_at: Date;
  procedure: string;
}
